apps/api/src/service/permissionService.ts

```diff
@@ -0,0 +1,114 @@
+ * This function grants a user access to the requested studies
+ * First, it looks for the associated PCGL user ID using the institutional email
+ * Then, it grants the study permissions using the approver's access token
+ * Returns the list of successfully granted study IDs and any failure messages
+ *
+ * @param param0
+ * @returns
+ */
+export const grantUserPermissions = async ({
+	institutionalEmail,
+	approverAccessToken,
+	requestedStudies,
+}: GrantUserPermissionsParams): Promise<GrantUserPermissionsResult> => {
+	const lookup = await lookupUserByEmail(institutionalEmail, approverAccessToken);
+
+	if (!lookup.success) {
+		return { success: false, failureMessages: [lookup.message] };
+	}
+
+	const failureMessages = [];
+	for (const studyId of requestedStudies ?? []) {
+		for (const userPcglId of lookup.data) {
+			const result = await addUserToStudyPermission(studyId, userPcglId, approverAccessToken);
+			if (!result.success) {
+				failureMessages.push(result.message);
+			}
+		}
+	}
+
+	return { success: failureMessages.length === 0, failureMessages };
+};
+
+/**
+ * Verifies that the applicant and all collaborators associated with an application
+ * have active user accounts and returns their user IDs.
+ * @param applicationId
+ * @param approverAccessToken
+ * @returns
+ */
+export const verifyApplicationUserAccounts = async (
+	applicationId: number,
+	approverAccessToken: string,
+): AsyncResult<{ userIds: string[] }, 'SYSTEM_ERROR' | 'APPLICATION_USERS_NOT_FOUND'> => {
+	const database = getDbInstance();
+
+	const applicationService: ApplicationService = applicationSvc(database);
+	const collaboratorsService = await collaboratorsSvc(database);
+
+	const emailLookupFailures: string[] = [];
+	const verifiedUserIds: string[] = [];
+
+	// Check if applicant has an active account
+	const applicationContents = await applicationService.getApplicationWithContents({ id: applicationId });
+	if (!applicationContents.success) {
+		return failure('SYSTEM_ERROR', 'Unable to retrieve application contents.');
+	}
+
+	const applicantEmail = applicationContents.data.contents?.applicant_institutional_email || '';
+
+	const applicantLookup = await lookupUserByEmail(applicantEmail, approverAccessToken);
+	if (!applicantLookup.success) {
+		emailLookupFailures.push(applicantEmail);
+	} else {
+		verifiedUserIds.push(...applicantLookup.data);
+	}
+
+	// Check if collaborators have active accounts
+	const collaboratorResp = await collaboratorsService.listCollaborators(applicationId);
+	if (!collaboratorResp.success) {
+		return failure('SYSTEM_ERROR', `Unable to retrieve collaborators: ${collaboratorResp.message}`);
+	}
+	for (const collaborator of collaboratorResp.data) {
+		const collabEmail = collaborator.institutional_email || '';
+		const collabLookup = await lookupUserByEmail(collabEmail, approverAccessToken);
+		if (!collabLookup.success) {
+			emailLookupFailures.push(collabEmail);
+		} else {
+			verifiedUserIds.push(...collabLookup.data);
+		}
```

Comment:
No reason to do negative boolean case here.
```suggestion
		if (collabLookup.success) {
			verifiedUserIds.push(...collabLookup.data);
		} else {
			emailLookupFailures.push(collabEmail);
		}
```